src/app/features/home/details/actions/actions.page.ts

```diff
@@ -51,39 +51,20 @@ export class ActionsPage {
         concatMap(
           ([params, token, id]) =>
             new Promise<void>(resolve => {
-              this.alertController
-                .create({
-                  header: action.title_text,
-                  message: action.description_text,
-                  inputs: params.map(
-                    param =>
-                      ({
-                        name: param.name_text,
-                        label: param.display_text_text,
-                        type: param.type_text,
-                        placeholder: param.placeholder_text,
-                        value: param.default_values_list_text[0],
-                        disabled: !param.user_input_boolean,
-                      } as AlertInput)
-                  ),
-                  buttons: [
-                    {
-                      text: this.translocoService.translate('cancel'),
-                      role: 'cancel',
-                    },
-                    {
-                      text: this.translocoService.translate('ok'),
-                      handler: value => {
-                        const body = { ...value, token: token, cid: id };
-                        return this.sendAction(action, body);
-                      },
-                    },
-                  ],
-                })
-                .then(alert => {
-                  alert.present();
-                  resolve();
-                });
+              const dialogRef = this.dialog.open(ActionsDialogComponent, {
+                disableClose: true,
+                data: {
+                  action: action,
+                  params: params,
+                },
+              });
+              dialogRef.afterClosed().subscribe(data => {
+                if (data !== undefined) {
+                  const body = { ...data, token: token, cid: id };
+                  return this.sendAction(action, body);
+                }
+              });
+              resolve();
```

Comment:
Reasons to make changes:
1. Nested subscription is anti-pattern
2. Wrapping the anonymous function with Promise here is redundant
3. Subscriptions never complete if not leaving this page

The thing we want to do here can be break down into:
1. Open a dialog and get a `dialogRef` object (synchronous)
2. Subscribe to `dialogRef.afterClose()` observable and perform actions (asynchronous)
3. Call `sendAction()` (synchronous)

After breaking down, we could see that there is only one asynchronous function call here, so this will be what we return in the callback for `concatMap` operator. If we return the observable, the observable will be subscribed by the outmost `subscribe()`, so there's no need to add another subscribe inside. Here we got 1 and 2 covered. And we can filter out `undefined` with the helper operator `isNonNullable`.
To perform a synchronous side effect in RxJS pipe, we could use `tap` operator. Placing the `isNonNullable` and `tap` operator in the pipe of inner observable or to the position after `concatMap` does not matter in this case.

Also, this subscription never completes unless completed by `untilDestroyed` by leaving the page. Add a `first()` operator at the start of the pipe will ensure this subscription completes after emitting once. Where to place `first()` does matter, since the subscription won't emit value when cancelling (because `undefined` will be filtered by `isNonNullable()`), so placing `first()` at the end of the pipe will make the subscription still never complete when action is cancelled.

```
        first(),
        concatMap(([params, token, id]) => {
          const dialogRef = this.dialog.open<ActionsDialogComponent>(
            ActionsDialogComponent,
            {
              disableClose: true,
              data: {
                action: action,
                params: params,
              },
            }
          );
          return dialogRef.afterClosed().pipe(
            isNonNullable(),
            tap(data =>
              this.sendAction(action, { ...data, token: token, cid: id })
            )
          );
        }),
```